### PR TITLE
NADTelnet: Fix communication function

### PR DESCRIPTION
**Pull request by andreas-amlabs**

### CODE REVIEW

File: nad_receiver/nad_transport.py [hunk 1]
```diff
@@ -81,8 +81,24 @@ def _open_connection(self) -> None:
...
+            msg = self.telnet.read_until(b"\r", self.timeout)
+            _LOGGER.debug("Read response: %s", str(msg))
+        except EOFError as cc:
+            # Connection closed
+            _LOGGER.debug("Connection closed: %s", cc)
+            self.telnet = None
+            msg = b""
+        except UnicodeError as ue:
+            _LOGGER.debug("Unicode error: %s", ue)
+            msg = b""
```
[the remark this answers is not in the document]
Comment (gladhorn) [reply nad_receiver/nad_transport.py:102]: And I'd be curious if you did ever see this one. That would mean NAD break their own protocol (at least all pdfs I found emphasized that it's ascii only.

File: nad_receiver/nad_transport.py [hunk 2]
```diff
@@ -81,8 +81,24 @@ def _open_connection(self) -> None:
 
     def communicate(self, cmd: str) -> str:
         self._open_connection()
-        assert self.telnet
+        if not self.telnet:
+            return b""
```
[the remark this answers is not in the document]
Comment (andreas-amlabs) [reply nad_receiver/nad_transport.py:85]: Actaully msg is bytes, so the func is wrong
"Read until a given byte string"
I will fix

File: nad_receiver/nad_transport.py [hunk 3]
```diff
@@ -81,8 +81,24 @@ def _open_connection(self) -> None:
...
+
+        try:
+            #_LOGGER.debug("Sending command: %s", cmd)
```
[the remark this answers is not in the document]
Comment (andreas-amlabs) [reply nad_receiver/nad_transport.py:88]: Sry, thats on me. Should be enabled, my mistake when re-enabling

File: nad_receiver/nad_transport.py [hunk 4]
```diff
@@ -81,8 +81,24 @@ def _open_connection(self) -> None:
...
+
+            # Notice NAD response to command ends with \r and starts with \n
+            # E.g. b'\nMain.Power=On\r'
+            msg = self.telnet.read_until(b"\r", self.timeout)
+            _LOGGER.debug("Read response: %s", str(msg))
+        except EOFError as cc:
+            # Connection closed
+            _LOGGER.debug("Connection closed: %s", cc)
+            self.telnet = None
+            msg = b""
```
[the remark this answers is not in the document]
Comment (andreas-amlabs) [reply nad_receiver/nad_transport.py:99]: Ok. I dont care

File: nad_receiver/nad_transport.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (andreas-amlabs) [reply nad_receiver/nad_transport.py:102]: Also catches the write, so Yes could happen ! And I have seen, but do not remember why

File: nad_receiver/nad_transport.py [hunk 5]
```diff
@@ -81,8 +81,24 @@ def _open_connection(self) -> None:
 
     def communicate(self, cmd: str) -> str:
         self._open_connection()
-        assert self.telnet
+        if not self.telnet:
+            return b""
+
+        try:
+            #_LOGGER.debug("Sending command: %s", cmd)
+            self.telnet.write(f"\n{cmd}\r".encode())
```
[the remark this answers is not in the document]
Comment (andreas-amlabs) [reply nad_receiver/nad_transport.py:89]: I write as I read. The read output from the nad matches this. Dunno which is actually correct, probably both. You want me to test the \r ?

[the remark this answers is not in the document]
Comment (andreas-amlabs) [reply nad_receiver/nad_transport.py:89]: If so, why ?

File: nad_receiver/nad_transport.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (andreas-amlabs) [reply nad_receiver/nad_transport.py:85]: But that wont match "super type NadTransport". Sigh

[the remark this answers is not in the document]
Comment (gladhorn) [reply nad_receiver/nad_transport.py:85]: Communicate takes a string and returns a string. Internally it encodes to bytes and decodes the answer again. Just changing to `return ""` fixes things cleanly.

File: nad_receiver/nad_transport.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (gladhorn) [reply nad_receiver/nad_transport.py:102]: OK, then we just take it in.

**Comment by andreas-amlabs** [comment]
You want any changes to this PR ?

**Pull request merged**